api/admin/controller/collection_settings.py

```diff
@@ -0,0 +1,266 @@
+            collections=collections,
+            protocols=self.protocols,
+        )
+
+    def collection_to_dict(self, collection_object):
+        return dict(
+            id=collection_object.id,
+            name=collection_object.name,
+            protocol=collection_object.protocol,
+            parent_id=collection_object.parent_id,
+            settings=dict(external_account_id=collection_object.external_account_id),
+        )
+
+    def load_libraries(self, collection_dict, collection_object, user, protocol):
+        libraries = []
+        for library in collection_object.libraries:
+            if not user or not user.is_librarian(library):
+                continue
+            libraries.append(self._get_integration_library_info(
+                    collection_object.external_integration, library, protocol))
+
+        collection_dict['libraries'] = libraries
+
+    def load_settings(self, protocol_settings, collection_dict, collection_object):
+        for protocol_setting in protocol_settings:
+            key = protocol_setting.get("key")
+            if key not in collection_dict.get("settings"):
+                if key == 'mirror_integration_id':
+                    value = collection_object.mirror_integration_id or self.NO_MIRROR_INTEGRATION
+                elif protocol_setting.get("type") == "list":
+                    value = collection_object.external_integration.setting(key).json_value
+                else:
+                    value = collection_object.external_integration.setting(key).value
+                collection_dict["settings"][key] = value
+
+    def find_protocol_class(self, collection_object):
+        protocolClassFound = [p for p in self.PROVIDER_APIS if p.NAME == collection_object.protocol]
+        if len(protocolClassFound) == 1:
+            return protocolClassFound
+
+
+    # POST
+    def process_post(self):
+        self.require_system_admin()
+        is_new = False
+        collection = None
+
+        name = flask.request.form.get("name")
+        protocol = flask.request.form.get("protocol")
+        fields = {"name": name, "protocol": protocol}
+        id = flask.request.form.get("id")
+        if id:
+            collection = get_one(self._db, Collection, id=id)
+            fields["collection"] = collection
+
+        error = self.validate_form_fields(is_new, **fields)
+        if error:
+            return error
+
+        if protocol and not collection:
+            collection, is_new = get_one_or_create(self._db, Collection, name=name)
+            if not is_new:
+                self._db.rollback()
+                return COLLECTION_NAME_ALREADY_IN_USE
+            collection.create_external_integration(fields.get("protocol"))
+
+        collection.name = name
+        [protocol] = [p for p in self.protocols if p.get("name") == protocol]
+
+        settings = self.validate_parent(protocol, collection)
+        if isinstance(settings, ProblemDetail):
+            return settings
```

Comment:
I think it would be good to move the various calls to `self._db.rollback()` out of the helper methods such as `validate_parent` and put them right before the `return` calls.

The reasoning is that `do_post` is the method that changed the database in the first place, by calling `get_one_or_create`. If it turns out that the database shouldn't have been modified, it's the responsibility of `do_post` to roll back the changes. Then a method like `validate_parent` can just focus on checking whether the incoming data is okay.